### DISCUSSION TITLE
Eventlet Fails Where Flask Development Server Works Fine

### DISCUSSION THREAD

**Discussion by Anti-Distinctlyminty** [miguelgrinberg/Flask-SocketIO#1790] [Q&A] [answered]
I am experiencing an issue whereby when running in the Flask development server everything works perfectly fine, but if I use eventlet, messages are not received and the socket seems to time out, disconnect and reconnect.

Here's the gist of my server code (I've omitted the code invoking the `socketio.emit`)

```python
# __init__.py

import simplejson

db = flask_sqlalchemy.SQLAlchemy()
api = flask_smorest.Api()
socketio = flask_socketio.SocketIO(json=simplejson, logger=True, engineio_logger=True)
```

```python
# app.py

import flask_cors
import flask

from . import (api, db, socketio)

def get_app():

    app = flask.Flask(__name__)
    app.config.from_mapping(...)
    app.app_context().push()   

    flask_cors.CORS(app)

    # flask_sqlalchemy
    db.init_app(app)
    db.Model.metadata.reflect(bind=db.engine, extend_existing=True)

    # flask_smorest
    api.init_app(app)
    # omitted register blueprints

    #flask_socketio
    socketio.init_app(app)

    return app

def run(port):
    app = get_app()
    socketio.run(app, host="0.0.0.0", port=port)

if __name__ == "__main__":
    sys.exit(run())
```

And on the client side

```python 
class Listen():
    def __init__(self)
        self.sio = socketio.Client(logger=True)

    def run(self):
        @self.sio.event
        def connect():
            log.info("PostgresListen socket connection established")

        @self.sio.event
        def disconnect():
            log.info("PostgresListen socket disconnected")

        @self.sio.on("projects")
        def onProjectChanged(data):
            log.info("PosgresListen project changed")
            log.debug(data)
            self.postgresEvent.emit(data)

        self.sio.connect(nymus.socketAddress)
```
When running with eventlet installed, when the server broadcasts an event I get a pause, then `PostgresListen socket disconnected` followed by `PostgresListen socket connection established`

Is there something about this setup that wont work with eventlet?

**Comment by miguelgrinberg** [comment, marked as answer]
Are you familiar with using asynchronous frameworks and green threads? If not, I think you shouldn't use eventlet. There are many things that need to be done differently, and many libraries that are incompatible.

**Reply by Anti-Distinctlyminty** [reply]
I'm not. I was originally using waitress as the production server, but the flask-socketio docs mentioned that eventlet was the simplest to implement. Do you have a recommendation for my situation?

**Reply by miguelgrinberg** [reply]
You can use Gunicorn multi-threaded: https://flask-socketio.readthedocs.io/en/latest/deployment.html#gunicorn-web-server. This is the third option described in this section of the docs.

You didn't say, but if you are on Windows, you don't have great options, as Gunicorn is incompatible. In that case my only suggestion is that you use WSL, where Gunicorn can be used.

**Reply by Anti-Distinctlyminty** [reply]
Yes I'm trying to keep things cross platform where possible, so Windows is a consideration. One thing I noted is that I just tried replacing `socketio.run(app, host="0.0.0.0", port=port)` with `waitress.serve(app, host="0.0.0.0", port=port)` and it seems to work as with the Flask development server. If this is not an obviously wrong thing to do I could carry on. the only reason I didn't try this before was because waitress wasn't listed in the deployment section of the docs.

**Reply by miguelgrinberg** [reply]
I don't think waitress is compatible. It will work for long-polling requests, but I don't think it supports WebSocket connections, which is kind of the main point of using Socket.IO.

**Reply by Anti-Distinctlyminty** [reply]
My use case is that I'm only trying to broadcast events from the server and have clients receive the data that is pushed, not communicate anything back (I have a REST APi for that). Am I perhaps using the wrong tool for this job? Flask-SocketIO seemed to do this and be easy to implement.

**Reply by miguelgrinberg** [reply]
There are many ways to do what you want, Socket.IO is one of them, so it is right for the job. I wouldn't say that it is easy to implement. None of the options that allow the server to push data to clients are easy to implement.

**Reply by Anti-Distinctlyminty** [reply]
In which case I shall look into Gunicorn on WSL as you suggest. Thank you Miguel.
